src/vs/workbench/contrib/chat/browser/chatSessions.contribution.ts

```diff
@@ -833,58 +832,29 @@ export class ChatSessionsService extends Disposable implements IChatSessionsServ
 		};
 	}
 
-	public registerModelProgressListener(model: IChatModel, callback: () => void): void {
-		// Prevent duplicate registrations for the same model
-		if (this._registeredModels.has(model)) {
-			return;
-		}
-		this._registeredModels.add(model);
-
-		// Helper function to register listeners for a request
-		const registerRequestListeners = (request: IChatRequestModel) => {
-			if (!request.response || this._registeredRequestIds.has(request.id)) {
-				return;
-			}
-
-			this._registeredRequestIds.add(request.id);
-
-			this._register(request.response.onDidChange(() => {
-				callback();
-			}));
-
-			// Track tool invocation state changes
-			const responseParts = request.response.response.value;
-			responseParts.forEach((part: IChatProgressResponseContent) => {
-				if (part.kind === 'toolInvocation') {
-					const toolInvocation = part as IChatToolInvocation;
-					// Use autorun to listen for state changes
-					this._register(autorunSelfDisposable(reader => {
-						const state = toolInvocation.state.read(reader);
+	public registerModelProgressListener(models: IChatModel[], type: string, emitter: Emitter<void>): IDisposable {
+		const disposables = new DisposableStore();
 
-						// Also track progress changes when executing
-						if (state.type === IChatToolInvocation.StateKind.Executing) {
-							state.progress.read(reader);
-						}
+		for (const model of models) {
+			if (model.sessionResource.scheme === type && !this._registeredModels.has(model)) {
+				this._registeredModels.add(model);
+				disposables.add(autorun(reader => {
+					const lastResponse = model.lastRequestObs.read(reader);
+					if (lastResponse?.response) {
+						disposables.add(lastResponse.response.onDidChange(() => {
+							emitter.fire();
+						}));
+					}
+				}));
 
-						callback();
-					}));
-				}
-			});
-		};
-		// Listen for response changes on all existing requests
-		const requests = model.getRequests();
-		requests.forEach(registerRequestListeners);
-
-		// Listen for new requests being added
-		this._register(model.onDidChange(() => {
-			const currentRequests = model.getRequests();
-			currentRequests.forEach(registerRequestListeners);
-		}));
+				disposables.add(model.onDidDispose(() => {
+					this._registeredModels.delete(model);
+					disposables.dispose();
```

Comment:
Seems incorrect, you have multiple sessions in the disposables but dispose all of them when any model gets disposed